### DISCUSSION TITLE
typescript  for states

### DISCUSSION THREAD

**Discussion by prbxr** [statelyai/xstate#2406] [Q&A] [answered]
Is it right that there is no way currently to type the state strings when writing the states config?
I tried many machine setups and it does not work.

**Comment by prbxr** [comment]
The only way that i see is to type states with enum 
```
[StateVariants.stateA]: { on: { onFocus: { target: StateVariants.stateB
```

**Reply by github0013** [reply]
https://xstate.js.org/docs/guides/ids.html#avoiding-strings
This might help?

**Reply by vzkhrv** [reply]
there is known issue with target getters, moreover [getters for targets will be removed in V5](https://github.com/statelyai/xstate/issues/678#issuecomment-599118532)

**Comment by prbxr** [comment]
Will this be improved in  next release?

**Comment by Andarist** [comment, marked as answer]
We will be working on improving the TypeScript experience - can't say when it will be released but this is very much on our minds, it just ain't a trivial problem for a library like XState. Stay tuned!

**Comment by vzkhrv** [comment]
@prbxr @Andarist I have the same issue and the only one workaround that i found instead of using maps or enums is to extend MachineConfig type. Since xstate doesn't specify error on wrong transition target neither in runtime nor in compile time, i found this issue major.

![Снимок экрана 2021-08-03 в 10 34 01](https://user-images.githubusercontent.com/8685395/127957605-a8492b39-798e-4bac-85a3-9f41941b523f.png)

![Снимок экрана 2021-08-03 в 10 34 41](https://user-images.githubusercontent.com/8685395/127957613-5fd9c363-e201-4db5-8e96-31a4da1eabfd.png)

Working demo: https://codesandbox.io/s/nice-engelbart-osqvc?file=/src/index.ts

Looks like

```js
import {
  createMachine,
  interpret,
  TransitionConfig,
  MachineConfig,
  StatesConfig,
  DoneInvokeEvent,
  SingleOrArray
} from "xstate";

const StateNames = ["IDLE", "OPENED", "CLOSED"] as const;

// machine config base types
type StateTypes = typeof StateNames[number];

type TContext = {};

type TStates = {
  states: Record<typeof StateNames[number], {}>;
};

type TEvents = { type: "open" } | { type: "close" };
//

type TransitionConfigExtracted<T> = TransitionConfig<
  TContext,
  Extract<TEvents, { type: T }>
>;

// u are able to describe every transition type one-by-one,
// if u need to specify targets separately
type TransitionConfigExtended = {
  [N in TEvents["type"]]?: SingleOrArray<
    { target: StateTypes } & TransitionConfigExtracted<N>
  >;
};

type StateConfigExtended = {
  on?: TransitionConfigExtended;
  invoke?: {
    onDone: SingleOrArray<
      { target: StateTypes } & TransitionConfig<TContext, DoneInvokeEvent<any>>
    >;
    onError: SingleOrArray<
      { target: StateTypes } & TransitionConfig<TContext, DoneInvokeEvent<any>>
    >;
  };
};

type MachineConfigExtended = MachineConfig<TContext, TStates, TEvents> & {
  states: StatesConfig<TContext, TStates, TEvents> &
    {
      // same for states
      [K in StateTypes]: StateConfigExtended;
    };
};

const config: MachineConfigExtended = {
  id: "cloneResume",
  initial: "IDLE",
  context: {},
  states: {
    IDLE: {
      on: {
        open: { target: "OPENED" },
        close: { target: "CLOSED" }
      }
    },
    OPENED: {
      on: {
        close: { target: "CLOSED" }
      }
    },
    CLOSED: {
      on: {
        open: { target: "OPENED" }
      }
    }
  }
};

const machine = createMachine(config);

// Edit your service(s) here
const service = interpret(machine).onTransition((state) => {
  console.log(state.value);
});

service.start();

setTimeout(() => {
  service.send({ type: "open" });
}, 1000);

setTimeout(() => {
  service.send({ type: "close" });
}, 2000);
```

**Comment by NamanRajputXXI** [comment]
GlobalState.tsx

import React, {  createContext } from "react";
import { useInterpret } from "@xstate/react";
import { gazeMachine } from "../machines/navigationMachine";
import {  InterpreterFrom } from "xstate";

export const GlobalStateContext = createContext({
  gazeService: {} as InterpreterFrom<typeof gazeMachine>,
});

export const GlobalStateProvider = (props: any) => {
  const gazeService = useInterpret(gazeMachine);

  return (
    <GlobalStateContext.Provider value={{ gazeService }}>
      {props.children}
    </GlobalStateContext.Provider>
  );
};

MinimalUi.tsx

import React, { useContext } from "react";
import { View, Text, TouchableOpacity } from "react-native";
import AddButton from "./AddButton";
import RandomApp from "./RandomApp";
import WidgetShell from "./WidgetShell";
import { GlobalStateContext } from "./GlobalState";
import {  useSelector } from "@xstate/react";


const passengerDetector = (state) => {
  return state.matches("passenger_detected");
};


const MinimalUi = (props: any) => {
  const globalServices = useContext(GlobalStateContext);

  const isPassengerIn = useSelector(
    globalServices.gazeService,
    passengerDetector
  );
  const { send } = globalServices.gazeService;

  return (
    <View className="flex gap-[15px]">
      <View className="h-[70%]  bg-white"></View>
      <View className="h-[180px] justify-center items-center flex flex-row gap-5  ">
        <WidgetShell />
        <WidgetShell />

        <AddButton />
        <TouchableOpacity
          onPress={() => {
            send("PASSENGER_DETECTED");
            console.log("hello");
          }}
        >
          <Text className="text-4xl font-normal text-white cursor-pointer">
            add app
          </Text>
        </TouchableOpacity>
      </View>
    </View>
  );
};

export default MinimalUi;
  I got an error undefined is not an object (evaluating state.matches)
i dont know what should be type of  this


const passengerDetector = (state) => {
  return state.matches("passenger_detected");
}; in thiss line i am getting error

can anyone help me i am stuck

**Reply by with-heart** [reply]
You can use `StateFrom` for this:

`state: StateFrom<typeof yourMachine>`

**Comment by NamanRajputXXI** [comment]
@with-heart  is this is a right approach of using  xstate machine with types

import { createMachine } from "xstate";

 export interface deviceContext {
  volume: number;
  gaze_timeout_s: number;
}
type Event =
  | { type: "PASSENGER_DETECTED"; value: "idle" }
  | {
      type: "PASSENGER_LEFT";
      value: "idle";
    }
  | {
      type: "GAZE";
      value: "in_gaze";
    }
  | {
      type: "GAZE_REMOVED";
      value: "idle";
    };
    

export const gazeMachine = createMachine<deviceContext, Event>({
  predictableActionArguments: true,
  id: "gaze",
  initial: "idle",

  context: {
    volume: 50,
    gaze_timeout_s: 1.0,
  },

  states: {
    idle: {
      on: {
        PASSENGER_DETECTED: "passenger_present",
      },
    },
    passenger_present: {
      initial: "idle",
      on: {
        PASSENGER_LEFT: "idle",
      },
      states: {
        idle: {
          on: {
            GAZE: "in_gaze",
          },
          after: {
            4000: {
              target: "attract_gaze",
            },
          },
        },
        attract_gaze: {
          entry: ["refresh_playlist"],
          on: {
            GAZE: "in_gaze",
          },
          after: {
            3000: "hyper_attract_gaze",
          },
        },
        hyper_attract_gaze: {
          on: {
            GAZE: "in_gaze",
          },
          after: {
            5000: "flip_channel",
          },
        },
        flip_channel: {
          on: {
            GAZE: "in_gaze",
          },
          after: {
            15000: "fail",
          },
        },
        fail: {
          on: {
            GAZE: "flip_channel",
          },
        },
        in_gaze: {
          on: {
            GAZE_REMOVED: "idle",
          },
          initial: "idle",
          states: {
            idle: {
              after: {
                3000: "cg1",
              },
            },
            cg1: {
              after: {
                5000: "cg2",
              },
            },
            cg2: {
              after: {
                8000: "cg3",
              },
            },
            cg3: {
              after: {
                1000: "offer_gift_coupon",
              },
            },
            offer_gift_coupon: {},
          },
        },
      },
    },
  },
});
